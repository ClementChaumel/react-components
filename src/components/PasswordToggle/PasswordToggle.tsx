import classNames from "classnames";
import React, { useEffect, useRef, useState } from "react";
import type { InputHTMLAttributes, ReactNode } from "react";

import Button from "../Button";
import Field from "../Field";
import Label from "../Label";

import type { ClassName, PropsWithSpread } from "types";

/**
 * The props for the Input component.
 */
export type Props = PropsWithSpread<
  {
    /**
     * The content for caution validation.
     */
    caution?: ReactNode;
    /**
     * Optional class(es) to pass to the input element.
     */
    className?: ClassName;
    /**
     * The content for error validation.
     */
    error?: ReactNode;
    /**
     * Help text to show below the field.
     */
    help?: ReactNode;
    /**
     * The id of the input.
     */
    id?: string;
    /**
     * The label for the field.
     */
    label?: ReactNode;
    /**
     * Whether the field is read only.
     */
    readOnly?: boolean;
    /**
     * Whether the field is required.
     */
    required?: boolean;
    /**
     * The content for success validation.
     */
    success?: ReactNode;
    /**
     * Whether to focus on the input on initial render.
     */
    takeFocus?: boolean;
    /**
     * Optional class(es) to pass to the wrapping Field component
     */
    wrapperClassName?: string;
  },
  InputHTMLAttributes<HTMLInputElement>
>;

const PasswordToggle = ({
  caution,
  className,
  error,
  help,
  id,
  label,
  readOnly,
  required,
  success,
  takeFocus,
  wrapperClassName,
  ...inputProps
}: Props): JSX.Element => {
  const [fieldType, setFieldType] = useState("password");
  const isPassword = fieldType === "password";
  const inputRef = useRef(null);

  const togglePassword = () => {
    if (isPassword) {
      setFieldType("text");
    } else {
      setFieldType("password");
    }
  };

  useEffect(() => {
    if (takeFocus) {
      inputRef.current.focus();
    }
  }, [takeFocus]);

  return (
    <Field
      caution={caution}
      className={wrapperClassName}
      error={error}
      help={help}
      required={required}
      success={success}
    >
      <div className="p-form-password-toggle">
        <Label forId={id} required={required}>
          {label}
        </Label>
        <Button
          appearance="base"
          className="u-no-margin--bottom"
          hasIcon
          aria-controls={id}
          aria-live="polite"
          onClick={() => togglePassword()}
        >
          <span className="p-form-password-toggle__label">
            {isPassword ? "Show" : "Hide"}
          </span>
          <i className={isPassword ? "p-icon--show" : "p-icon--hide"}></i>
        </Button>
      </div>
      <input
        className={classNames("p-form-validation__input", className)}
        id={id}
        ref={inputRef}
        type={fieldType}
        readOnly={readOnly}
        {...inputProps}
      />
    </Field>
  );
};

export default PasswordToggle;
