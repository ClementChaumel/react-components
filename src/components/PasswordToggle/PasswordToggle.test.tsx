import { mount, shallow } from "enzyme";
import React from "react";

import PasswordToggle from "./PasswordToggle";

describe("PasswordToggle ", () => {
  it("renders", () => {
    const wrapper = shallow(<PasswordToggle id="test-id" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("can add additional classes", () => {
    const wrapper = shallow(
      <PasswordToggle id="test-id" className="extra-class" />
    );
    const className = wrapper.find("input").prop("className");
    expect(className.includes("p-form-validation__input")).toBe(true);
    expect(className.includes("extra-class")).toBe(true);
  });

  it("can take focus on first render", () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    const wrapper = mount(<PasswordToggle id="test-id" takeFocus />, {
      attachTo: container,
    });
    expect(wrapper.find("input").getDOMNode()).toBe(document.activeElement);
  });

  it("toggles the visability of the password whent the button is clicked", () => {
    const wrapper = mount(<PasswordToggle id="test-id" />);
    expect(wrapper.find("input#test-id").prop("type")).toBe("password");
    wrapper.find("button[aria-controls='test-id']").simulate("click");
    expect(wrapper.find("input#test-id").prop("type")).toBe("text");
  });
});
