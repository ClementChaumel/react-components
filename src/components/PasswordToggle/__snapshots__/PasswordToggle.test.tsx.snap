// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordToggle  renders 1`] = `
<Field>
  <div
    className="p-form-password-toggle"
  >
    <Label
      forId="test-id"
    />
    <Button
      appearance="base"
      aria-controls="test-id"
      aria-live="polite"
      className="u-no-margin--bottom"
      hasIcon={true}
      onClick={[Function]}
    >
      <span
        className="p-form-password-toggle__label"
      >
        Show
      </span>
      <i
        className="p-icon--show"
      />
    </Button>
  </div>
  <input
    className="p-form-validation__input"
    id="test-id"
    type="password"
  />
</Field>
`;
